File: hackerrank/problem solving/algorithms/2 - implementation/3 - number line jumps/number_line_jumps.ts
// https://www.hackerrank.com/challenges/kangaroo/problem?isFullScreen=true

function main() {
  const initialState: number[] = readNums();
  console.log(kangaroo(initialState) ? 'YES' : 'NO');
}

function readNums(): number[] {
  return readLine().split(' ').map(Number);
}

// T: O(1)
// S: O(1) extra space
function kangaroo(initialState: number[]): boolean {
  const [x1, v1, x2, v2]: number[] = initialState;

  if (v1 === v2) return x1 === x2;

  const distanceDiff = x2 - x1;
  const velocityDiff = v1 - v2;
  return distanceDiff * velocityDiff >= 0 && distanceDiff % velocityDiff === 0;
}

//////////////////////////////////////////////////

import readline = require('readline');

const rl = readline.createInterface({ input: process.stdin });
const inputLines: string[] = [];
rl.on('line', (line: string) => inputLines.push(line));
rl.on('close', main);

function readLine(): string {
  return inputLines.shift()!;
}

File: hackerrank/problem solving/algorithms/2 - implementation/3 - number line jumps/number_line_jumps.py
# https://www.hackerrank.com/challenges/kangaroo/problem?isFullScreen=true

def main():
    initial_state = read_numbers()
    print("YES" if kangaroo(initial_state) else "NO")


def read_numbers():
    return list(map(int, input().split()))


# T: O(1)
# S: O(1) extra space
def kangaroo(initial_state):
    x1, v1, x2, v2 = initial_state

    if v1 == v2:
        return x1 == x2

    distance_diff = x2 - x1
    velocity_diff = v1 - v2
    return distance_diff * velocity_diff >= 0 and distance_diff % velocity_diff == 0


if __name__ == '__main__':
    main()
